portal/common/lib/rpc_selector.ts

```diff
@@ -0,0 +1,159 @@
+    }
+
+    // General method to call clients and return the first successful response.
+    private async invokeWithFailover<T>(methodName: string, args: any[]): Promise<T> {
+        if (this.clients.length === 0) {
+            throw new Error("No available clients to handle the request.");
+        }
+
+        const isNoSelectedClient = !this.selectedClient;
+        if (isNoSelectedClient) {
+            return await this.callFallbackClients<T>(methodName, args);
+        }
+
+        try {
+            return await this.callSelectedClient<T>(methodName, args);
+        } catch (error) {
+            console.error(`Selected client failed: ${error}`);
+            this.selectedClient = undefined;
+            return await this.callFallbackClients<T>(methodName, args);
+        }
+    }
+
+    // Attempt to call the method on the selected client with a timeout.
+    private async callSelectedClient<T>(methodName: string, args: any[]): Promise<T> {
+        const method = (this.selectedClient as any)[methodName] as Function;
+        if (!method) {
+            throw new Error(`Method ${methodName} not found on selected client`);
+        }
+        const timeoutDuration = 5000;
+        const timeoutPromise = new Promise<never>((_, reject) =>
+            setTimeout(() => reject(new Error("Request timed out")), timeoutDuration),
+        );
+
+        const result = await Promise.race([
+            method.apply(this.selectedClient, args),
+            timeoutPromise,
+        ]);
+
+        if (this.isValidResponse(result)) {
+            return result;
+        } else {
+            console.error("Invalid response from selected client");
+            throw new Error("Invalid response from selected client");
+        }
+    }
+
+    // Fallback to querying all clients using Promise.any.
+    private async callFallbackClients<T>(methodName: string, args: any[]): Promise<T> {
+        const clientPromises = this.clients.map((client) =>
+            new Promise<{ result: T; client: SuiClient }>(async (resolve, reject) => {
+                try {
+                    const method = (client as any)[methodName] as Function;
+                    if (!method) {
+                        reject(new Error(`Method ${methodName} not found on client`));
+                        return;
+                    }
+                    const result = await method.apply(client, args);
+                    if (this.isValidResponse(result)) {
+                        resolve({ result, client });
+                    } else {
+                        reject(new Error("Invalid response"));
+                    }
+                } catch (error: any) {
+                    reject(error);
+                }
+            }),
+        );
+
+        try {
+            const { result, client } = await Promise.any(clientPromises);
+            // Update the selected client for future calls.
+            this.selectedClient = client;
+            return result;
```

Comment:
Praise: I like this! 🚀 

Question: what is the "lifetime" of a selected client, depending on the way the portal is deployed? Something like:
- service worker: until the service worker is refreshed
- server side on edge functions: every call has a different one?

Suggestion: log which client was selected, for easier testing?